File: pkg-r/R/citation-scan.R
# Require explicit blockquote lines so wrapped evidence fails closed instead
# of verifying only its first line.
parse_commons_citation <- function(body) {
  lines <- strsplit(body, "\n", fixed = TRUE)[[1]]
  quoted <- grepl("^> ?", lines)
  runs <- rle(quoted)
  if (sum(runs$values) != 1) {
    return(NULL)
  }
  quote <- paste(sub("^> ?", "", lines[quoted]), collapse = "\n")
  explanation <- trimws(paste(lines[!quoted], collapse = "\n"))
  list(explanation = explanation, quote = quote)
}

# Hold partial reserved tags between chunks so projection does not depend on
# stream chunk boundaries.
CITATION_OPEN <- "<commons-citation>"
CITATION_CLOSE <- "</commons-citation>"
ASIDE_OPEN <- "<shiny-aside"
ASIDE_CLOSE <- "</shiny-aside>"
ELEMENT_BODY_CAP <- 16384L

citation_scanner <- function(corpus = list()) {
  resolve <- function(parsed) {
    if (is.null(parsed)) {
      return(list(
        html = "",
        decision = list(quote = NA_character_, status = "malformed")
      ))
    }
    render_citation_aside(parsed$quote, parsed$explanation, corpus)
  }
  buf <- ""
  mode <- "text"
  at_line_start <- TRUE
  decisions <- list()
  out <- character(0)
  discard_close <- NULL

  emit <- function(text) {
    if (nzchar(text)) out[[length(out) + 1]] <<- text
  }

  # Tag anchoring follows the model's input, not the projected output.
  note_line_start <- function(original_text) {
    if (nzchar(original_text)) {
      at_line_start <<- endsWith(original_text, "\n")
    }
  }

  record <- function(decision) {
    decisions[[length(decisions) + 1]] <<- decision
  }

  begin_discard <- function(close_literal) {
    mode <<- "discard"
    discard_close <<- close_literal
  }

  step <- function() {
    if (mode == "text") {
      event <- find_text_event(buf, at_line_start)
      if (!is.null(event)) {
        prefix <- substr(buf, 1, event$pos - 1)
        literal <- substr(buf, event$pos, event$pos + event$len - 1L)
        emit(prefix)
        note_line_start(prefix)
        buf <<- substr(buf, event$pos + event$len, nchar(buf))
        if (identical(event$mode, "citation")) {
          mode <<- "citation"
        } else if (identical(event$mode, "aside")) {
          begin_discard(ASIDE_CLOSE)
        } else {
          note_line_start(literal)
        }
        return(TRUE)
      }
      holdback <- holdback_length(buf, at_line_start)
      flush_len <- nchar(buf) - holdback
      if (flush_len > 0) {
        flushed <- substr(buf, 1, flush_len)
        emit(flushed)
        note_line_start(flushed)
        buf <<- substr(buf, flush_len + 1, nchar(buf))
      }
      return(FALSE)
    }

    if (mode == "citation") {
      event <- find_reserved_event(buf)
      if (
        !is.null(event) &&
          identical(event$action, "close") &&
          identical(event$kind, "citation") &&
          (event$pos - 1L) <= ELEMENT_BODY_CAP
      ) {
        body <- substr(buf, 1, event$pos - 1L)
        buf <<- substr(buf, event$pos + event$len, nchar(buf))
        close_citation(body)
        mode <<- "text"
        at_line_start <<- FALSE
        return(TRUE)
      }
      if (!is.null(event)) {
        begin_discard(CITATION_CLOSE)
        return(TRUE)
      }
      if (confirmed_body_len(buf, CITATION_CLOSE) > ELEMENT_BODY_CAP) {
        begin_discard(CITATION_CLOSE)
        return(TRUE)
      }
      return(FALSE)
    }

    pos <- find_ci(buf, discard_close)
    if (!is.na(pos)) {
      consumed <- substr(buf, 1, pos + nchar(discard_close) - 1L)
      note_line_start(consumed)
      buf <<- substr(buf, pos + nchar(discard_close), nchar(buf))
      mode <<- "text"
      discard_close <<- NULL
      return(TRUE)
    }

    holdback <- longest_valid_suffix(
      buf,
      discard_close,
      function(start) TRUE
    )
    drop_len <- nchar(buf) - holdback
    if (drop_len > 0) {
      dropped <- substr(buf, 1, drop_len)
      note_line_start(dropped)
      buf <<- substr(buf, drop_len + 1L, nchar(buf))
    }
    FALSE
  }

  close_citation <- function(body) {
    result <- resolve(parse_commons_citation(body))
    emit(result$html)
    record(result$decision)
    invisible()
  }

  list(
    feed = function(chunk) {
      buf <<- paste0(buf, chunk)
      out <<- character(0)
      while (step()) {}
      paste(out, collapse = "")
    },
    finish = function() {
      if (mode == "text") {
        flushed <- buf
        buf <<- ""
        return(flushed)
      }
      # Never expose incomplete model-authored markup.
      buf <<- ""
      mode <<- "text"
      discard_close <<- NULL
      ""
    },
    decisions = function() decisions
  )
}

project_citation_text <- function(text, corpus) {
  s <- citation_scanner(corpus)
  out <- paste0(s$feed(text), s$finish())
  list(text = out, decisions = s$decisions())
}

find_text_event <- function(buf, at_line_start) {
  citation_pattern <- if (at_line_start) {
    "(?:^|(?<=\n))<commons-citation>"
  } else {
    "(?<=\n)<commons-citation>"
  }
  citation_pos <- regexpr(
    citation_pattern,
    buf,
    perl = TRUE,
    ignore.case = TRUE
  )
  aside_pos <- regexpr(tolower(ASIDE_OPEN), tolower(buf), fixed = TRUE)

  candidates <- list()
  if (citation_pos != -1) {
    candidates[[length(candidates) + 1]] <- list(
      pos = as.integer(citation_pos),
      len = nchar(CITATION_OPEN),
      mode = "citation"
    )
  }
  if (aside_pos != -1) {
    candidates[[length(candidates) + 1]] <- list(
      pos = as.integer(aside_pos),
      len = nchar(ASIDE_OPEN),
      mode = "aside"
    )
  }
  citation_close_pos <- find_ci(buf, CITATION_CLOSE)
  if (!is.na(citation_close_pos)) {
    candidates[[length(candidates) + 1]] <- list(
      pos = citation_close_pos,
      len = nchar(CITATION_CLOSE),
      mode = "drop"
    )
  }
  aside_close_pos <- find_ci(buf, ASIDE_CLOSE)
  if (!is.na(aside_close_pos)) {
    candidates[[length(candidates) + 1]] <- list(
      pos = aside_close_pos,
      len = nchar(ASIDE_CLOSE),
      mode = "drop"
    )
  }
  if (length(candidates) == 0) {
    return(NULL)
  }
  candidates[[which.min(vapply(candidates, function(c) c$pos, integer(1)))]]
}

find_ci <- function(buf, literal) {
  pos <- regexpr(tolower(literal), tolower(buf), fixed = TRUE)
  if (pos == -1) NA_integer_ else as.integer(pos)
}

find_reserved_event <- function(buf) {
  events <- list(
    list(literal = CITATION_OPEN, kind = "citation", action = "open"),
    list(literal = CITATION_CLOSE, kind = "citation", action = "close"),
    list(literal = ASIDE_OPEN, kind = "aside", action = "open"),
    list(literal = ASIDE_CLOSE, kind = "aside", action = "close")
  )
  for (i in seq_along(events)) {
    events[[i]]$pos <- find_ci(buf, events[[i]]$literal)
    events[[i]]$len <- nchar(events[[i]]$literal)
  }
  events <- Filter(function(event) !is.na(event$pos), events)
  if (length(events) == 0) {
    return(NULL)
  }
  events[[which.min(vapply(events, function(event) event$pos, integer(1)))]]
}

# Exclude a partial closing tag from the body cap so chunk boundaries cannot
# change whether a citation is accepted.
confirmed_body_len <- function(buf, close_literal) {
  holdback <- longest_valid_suffix(buf, close_literal, function(start) TRUE)
  nchar(buf) - holdback
}

holdback_length <- function(buf, at_line_start) {
  citation_anchor <- function(start) {
    if (start == 1) {
      at_line_start
    } else {
      identical(substr(buf, start - 1, start - 1), "\n")
    }
  }
  max(
    longest_valid_suffix(buf, CITATION_OPEN, citation_anchor),
    longest_valid_suffix(buf, ASIDE_OPEN, function(start) TRUE),
    longest_valid_suffix(buf, CITATION_CLOSE, function(start) TRUE),
    longest_valid_suffix(buf, ASIDE_CLOSE, function(start) TRUE)
  )
}

longest_valid_suffix <- function(buf, literal, anchor_ok) {
  n <- nchar(buf)
  max_len <- min(n, nchar(literal) - 1L)
  if (max_len < 1L) {
    return(0L)
  }
  for (len in max_len:1) {
    start <- n - len + 1L
    if (!anchor_ok(start)) {
      next
    }
    if (is_ci_prefix(substr(buf, start, n), literal, len)) {
      return(len)
    }
  }
  0L
}

is_ci_prefix <- function(suffix, literal, len) {
  identical(tolower(suffix), tolower(substr(literal, 1, len)))
}

File: pkg-py/src/commons/_citation_scan.py
"""Project a model's streamed text for display, holding reserved markup back.

The projection is a cross-language contract pinned by the ``citation_scan``
cases in ``tests/shared/citations.json``; change that fixture, not just this
file. ``pkg-r/R/citation-scan.R`` implements the same contract for R.

Reserved markup is anything the display layer treats as trusted: a
``<commons-citation>`` block, which becomes an aside only once its quote
verifies, and a ``<shiny-aside>``, which the model must never be able to author.
Output cannot depend on where a chunk boundary falls, so text that could still
grow into a reserved tag is held back until the next chunk decides it.
"""

from __future__ import annotations

import re
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal

from ._citations import (
    CitationDecision,
    CorpusEntry,
    citation_aside_html,
    match_citation,
    parse_commons_citation,
)

__all__ = ["CitationScanner"]

CITATION_OPEN = "<commons-citation>"
CITATION_CLOSE = "</commons-citation>"
# Attributes may follow, so the opening aside tag is matched without its bracket.
ASIDE_OPEN = "<shiny-aside"
ASIDE_CLOSE = "</shiny-aside>"

# A body this large is a runaway rather than a citation. Past it the element is
# abandoned, which is what keeps an unclosed tag from buffering without limit.
ELEMENT_BODY_CAP = 16384

_Mode = Literal["text", "citation", "discard"]
_Action = Literal["citation", "aside", "drop", "close"]

# Matched as patterns rather than by lowercasing the buffer, because casefolding
# can change a string's length and would shift every position that follows.
_LITERALS = {
    literal: re.compile(re.escape(literal), re.IGNORECASE)
    for literal in (CITATION_OPEN, CITATION_CLOSE, ASIDE_OPEN, ASIDE_CLOSE)
}
_PREFIXES = {
    literal: [
        re.compile(re.escape(literal[:length]) + r"\Z", re.IGNORECASE)
        for length in range(len(literal))
    ]
    for literal in _LITERALS
}
# The opening tag only counts at the start of a line, so prose that mentions it
# mid-sentence survives.
_CITATION_OPEN_ANCHORED = re.compile(
    r"(?<=\n)" + re.escape(CITATION_OPEN), re.IGNORECASE
)
_CITATION_OPEN_AT_START = re.compile(
    r"(?:^|(?<=\n))" + re.escape(CITATION_OPEN), re.IGNORECASE
)


@dataclass(frozen=True)
class _Event:
    """A reserved literal found in the buffer, and what to do with it."""

    pos: int
    length: int
    action: _Action


def _find(literal: str, buffer: str) -> re.Match[str] | None:
    return _LITERALS[literal].search(buffer)


def _held_back(buffer: str, literal: str, at_line_start: bool | None = None) -> int:
    """Length of the longest buffer suffix that could still grow into ``literal``.

    Pass ``at_line_start`` for a literal that only counts at the start of a
    line, so a tail that could never open an element is emitted rather than
    held; leave it None for the literals that count at any position.
    """
    for length in range(min(len(buffer), len(literal) - 1), 0, -1):
        start = len(buffer) - length
        if at_line_start is not None:
            anchored = at_line_start if start == 0 else buffer[start - 1] == "\n"
            if not anchored:
                continue
        if _PREFIXES[literal][length].match(buffer, start):
            return length
    return 0


class CitationScanner:
    """Rewrite verified citations and drop every other reserved element.

    Feed the model's chunks in and emit what is safe to display now. The
    scanner projects text for display and never rewrites what the caller
    stores, so the turn keeps the model's own words.
    """

    def __init__(
        self,
        corpus: Sequence[CorpusEntry] = (),
        *,
        max_element_bytes: int = ELEMENT_BODY_CAP,
    ) -> None:
        self._corpus = list(corpus)
        self._cap = max_element_bytes
        self._buffer = ""
        self._mode: _Mode = "text"
        self._at_line_start = True
        self._discard_close: str | None = None
        self._candidates: list[CitationDecision] = []
        self._out: list[str] = []

    def feed(self, chunk: str) -> str:
        """Consume a stream chunk and return the text safe to emit now."""
        self._buffer += chunk
        self._out = []
        while self._step():
            pass
        return "".join(self._out)

    def finish(self) -> str:
        """Flush held-back text at end of stream."""
        if self._mode == "text":
            flushed, self._buffer = self._buffer, ""
            return flushed
        # Incomplete model-authored markup must never reach the display.
        self._buffer = ""
        self._mode = "text"
        self._discard_close = None
        return ""

    @property
    def candidates(self) -> list[CitationDecision]:
        """Every candidate citation and its verdict, in the order they arrived."""
        return list(self._candidates)

    @property
    def any_verified(self) -> bool:
        """Whether any candidate's quote was found in the corpus."""
        return any(candidate.status == "accepted" for candidate in self._candidates)

    def _step(self) -> bool:
        """Advance once. True when the buffer moved and another step may apply."""
        if self._mode == "text":
            return self._step_text()
        if self._mode == "citation":
            return self._step_citation()
        return self._step_discard()

    def _step_text(self) -> bool:
        event = self._find_text_event()
        if event is not None:
            prefix = self._buffer[: event.pos]
            literal = self._buffer[event.pos : event.pos + event.length]
            self._emit(prefix)
            self._note_line_start(prefix)
            self._buffer = self._buffer[event.pos + event.length :]
            if event.action == "citation":
                self._mode = "citation"
            elif event.action == "aside":
                self._begin_discard(ASIDE_CLOSE)
            else:
                self._note_line_start(literal)
            return True

        flush_to = len(self._buffer) - self._holdback_length()
        if flush_to > 0:
            flushed = self._buffer[:flush_to]
            self._emit(flushed)
            self._note_line_start(flushed)
            self._buffer = self._buffer[flush_to:]
        return False

    def _step_citation(self) -> bool:
        event = self._find_reserved_event()
        if event is not None and event.action == "close" and event.pos <= self._cap:
            body = self._buffer[: event.pos]
            self._buffer = self._buffer[event.pos + event.length :]
            self._close_citation(body)
            self._mode = "text"
            self._at_line_start = False
            return True
        # Any other reserved literal inside the body means the element is not a
        # citation at all, so it is abandoned rather than parsed.
        if event is not None:
            self._begin_discard(CITATION_CLOSE)
            return True
        if self._confirmed_body_length() > self._cap:
            self._begin_discard(CITATION_CLOSE)
            return True
        return False

    def _step_discard(self) -> bool:
        assert self._discard_close is not None
        match = _find(self._discard_close, self._buffer)
        if match is not None:
            self._note_line_start(self._buffer[: match.end()])
            self._buffer = self._buffer[match.end() :]
            self._mode = "text"
            self._discard_close = None
            return True

        drop_to = len(self._buffer) - _held_back(self._buffer, self._discard_close)
        if drop_to > 0:
            self._note_line_start(self._buffer[:drop_to])
            self._buffer = self._buffer[drop_to:]
        return False

    def _find_text_event(self) -> _Event | None:
        pattern = (
            _CITATION_OPEN_AT_START if self._at_line_start else _CITATION_OPEN_ANCHORED
        )
        found: list[tuple[re.Match[str] | None, int, _Action]] = [
            (pattern.search(self._buffer), len(CITATION_OPEN), "citation"),
            (_find(ASIDE_OPEN, self._buffer), len(ASIDE_OPEN), "aside"),
            # A close tag with no open one is stray markup and is dropped, but
            # it leaves the text around it alone.
            (_find(CITATION_CLOSE, self._buffer), len(CITATION_CLOSE), "drop"),
            (_find(ASIDE_CLOSE, self._buffer), len(ASIDE_CLOSE), "drop"),
        ]
        events = [
            _Event(match.start(), length, action)
            for match, length, action in found
            if match is not None
        ]
        return min(events, key=lambda event: event.pos, default=None)

    def _find_reserved_event(self) -> _Event | None:
        found: list[tuple[re.Match[str] | None, int, _Action]] = [
            (_find(CITATION_OPEN, self._buffer), len(CITATION_OPEN), "drop"),
            (_find(CITATION_CLOSE, self._buffer), len(CITATION_CLOSE), "close"),
            (_find(ASIDE_OPEN, self._buffer), len(ASIDE_OPEN), "drop"),
            (_find(ASIDE_CLOSE, self._buffer), len(ASIDE_CLOSE), "drop"),
        ]
        events = [
            _Event(match.start(), length, action)
            for match, length, action in found
            if match is not None
        ]
        return min(events, key=lambda event: event.pos, default=None)

    def _confirmed_body_length(self) -> int:
        """Body length so far, less a partial close tag that may still complete.

        Counting the partial tag would let a chunk boundary push an in-range
        body over the cap.
        """
        return len(self._buffer) - _held_back(self._buffer, CITATION_CLOSE)

    def _holdback_length(self) -> int:
        return max(
            _held_back(self._buffer, CITATION_OPEN, self._at_line_start),
            _held_back(self._buffer, ASIDE_OPEN),
            _held_back(self._buffer, CITATION_CLOSE),
            _held_back(self._buffer, ASIDE_CLOSE),
        )

    def _close_citation(self, body: str) -> None:
        parsed = parse_commons_citation(body)
        if parsed is None:
            self._candidates.append(CitationDecision(quote=None, status="malformed"))
            return
        entry = match_citation(parsed.quote, self._corpus)
        if entry is None:
            self._candidates.append(
                CitationDecision(quote=parsed.quote, status="rejected")
            )
            return
        self._emit(
            citation_aside_html(
                parsed.quote, parsed.explanation, entry.label, entry.kind
            )
        )
        self._candidates.append(
            CitationDecision(
                quote=parsed.quote,
                status="accepted",
                label=entry.label,
                kind=entry.kind,
            )
        )

    def _begin_discard(self, close_literal: str) -> None:
        self._mode = "discard"
        self._discard_close = close_literal

    def _emit(self, text: str) -> None:
        if text:
            self._out.append(text)

    def _note_line_start(self, original: str) -> None:
        # Tag anchoring follows the model's own text, not the projected output.
        if original:
            self._at_line_start = original.endswith("\n")
